ee/institution/admin/saml.sql

```diff
@@ -2,47 +2,84 @@
 SELECT * FROM institutions WHERE id = $id;
 
 -- BLOCK insert_institution_saml_provider
-INSERT INTO saml_providers (
-  institution_id,
-  sso_login_url,
-  issuer,
-  certificate,
-  uid_attribute,
-  uin_attribute,
-  name_attribute,
-  public_key,
-  private_key
-) VALUES (
-  $institution_id,
-  $sso_login_url,
-  $issuer,
-  $certificate,
-  $uid_attribute,
-  $uin_attribute,
-  $name_attribute,
-  COALESCE($public_key, ''),
-  COALESCE($private_key, '')
-) ON CONFLICT (institution_id) DO UPDATE
-SET
-  sso_login_url = EXCLUDED.sso_login_url,
-  issuer = EXCLUDED.issuer,
-  certificate = EXCLUDED.certificate,
-  uid_attribute = EXCLUDED.uid_attribute,
-  uin_attribute = EXCLUDED.uin_attribute,
-  name_attribute = EXCLUDED.name_attribute,
-  public_key = COALESCE($public_key, saml_providers.public_key),
-  private_key = COALESCE($private_key, saml_providers.private_key);
+WITH existing_saml_provider AS (
+  SELECT * FROM saml_providers WHERE institution_id = $institution_id
+), upserted_saml_provider AS (
+  INSERT INTO saml_providers (
+    institution_id,
+    sso_login_url,
+    issuer,
+    certificate,
+    uid_attribute,
+    uin_attribute,
+    name_attribute,
+    public_key,
+    private_key
+  ) VALUES (
+    $institution_id,
+    $sso_login_url,
+    $issuer,
+    $certificate,
+    $uid_attribute,
+    $uin_attribute,
+    $name_attribute,
+    COALESCE($public_key, ''),
+    COALESCE($private_key, '')
+  ) ON CONFLICT (institution_id) DO UPDATE
+  SET
+    sso_login_url = EXCLUDED.sso_login_url,
+    issuer = EXCLUDED.issuer,
+    certificate = EXCLUDED.certificate,
+    uid_attribute = EXCLUDED.uid_attribute,
+    uin_attribute = EXCLUDED.uin_attribute,
+    name_attribute = EXCLUDED.name_attribute,
+    public_key = COALESCE($public_key, saml_providers.public_key),
+    private_key = COALESCE($private_key, saml_providers.private_key)
+  RETURNING *
+), audit_logs_upserted_saml_provider AS (
+  INSERT INTO audit_logs (
+    authn_user_id,
+    table_name,
+    action,
+    institution_id,
+    row_id,
+    old_state,
+    new_state
+  ) SELECT
+    $authn_user_id as authn_user_id,
+    'saml_providers' as table_name,
+    (CASE WHEN esp.id IS NULL THEN 'insert' ELSE 'update' END) as action,
+    $institution_id as institution_id,
+    usp.id AS row_id,
+    (to_jsonb(esp) || jsonb_build_object('private_key', 'REDACTED')) AS old_state,
```

Comment:
With my security hat on, I like the general policy of stripping sensitive info from logs. But with my customer support hat on, I'm less sure. Am I correct in thinking that the `private_key` is never exposed to the user (or outside the server)? If so, I'm not sure I see what the attack is that would make use of old private keys.

However, I also don't think that in practice we'll need to recover old private keys, so I'm not very worried one way or the other.

Reply:
> Am I correct in thinking that the private_key is never exposed to the user (or outside the server)?

Correct, yeah. I'm thinking forward to a future where, like in PT, we have user-facing audit logs pages, and I think even in that case we'd never want the possibility of exposing the private key. That said, yeah, I don't know what vulnerabilities we'd actually get if folks had our private key; I think at worst, if they were somehow able to intercept the redirect from the IDP, they'd be able to decrypt the IDP identity assertion?